js/store.js

```diff
@@ -112,49 +217,137 @@ const store = {
+      for (let i = 0; i < hostnameVariantsFirstParty.length; i++) {
+        if (this.firstPartyWhiteList
+          .hasOwnProperty(hostnameVariantsFirstParty[i])) {
+          // first party is in the whitelist
+          const index = this.firstPartyWhiteList[hostnameVariantsFirstParty[i]];
+          const hostnameVariantsThirdParty
+            = this.getHostnameVariants(thirdPartyFromRequest);
+          for (let j = 0; j < hostnameVariantsThirdParty.length; j++) {
+            if (this.thirdPartyWhiteList[index]
+              .includes(hostnameVariantsThirdParty[j])) {
+              return true;
+            }
+          }
+          return false;
+        }
+      }
+    }
+    return false;
   },
 
   setFirstParty(hostname, data) {
     if (!hostname) {
       throw new Error('setFirstParty requires a valid hostname argument');
     }
+    const isNewWebsite = this.isNewWebsite(hostname);
+
     this.setWebsite(hostname, data);
+
+    if (isNewWebsite) {
+      this.updateChild(this.outputWebsite(hostname, data));
+    }
   },
 
+  // @todo restructure storage to use data from capture for graph filtering
   setThirdParty(origin, target, data) {
-    let newThirdParty = false;
     if (!origin) {
       throw new Error('setThirdParty requires a valid origin argument');
     }
 
+    let isNewThirdParty = false;
+    let isNewFirstParty = false;
+
     const firstParty = this.getWebsite(origin);
     const thirdParty = this.getWebsite(target);
 
-    if (!('thirdPartyHostnames' in firstParty)) {
-      firstParty['thirdPartyHostnames'] = [];
-    }
-    if (!firstParty['thirdPartyHostnames'].includes(target)) {
-      firstParty['thirdPartyHostnames'].push(target);
-      newThirdParty = true;
-    }
-    if (!('firstPartyHostnames' in firstParty)) {
+    // add link in third party if it doesn't exist yet
+    if (!('firstPartyHostnames' in thirdParty)) {
       thirdParty['firstPartyHostnames'] = [];
     }
     if (!thirdParty['firstPartyHostnames'].includes(origin)) {
       thirdParty['firstPartyHostnames'].push(origin);
-      newThirdParty = true;
+      isNewFirstParty = true;
+    }
+
+    // add link in first party if it doesn't exist yet
+    // and the third party is visible (i.e. not whitelisted)
+    if (!('thirdPartyHostnames' in  firstParty)
+      || !firstParty['thirdPartyHostnames'].includes(target)) {
+      if (!('isVisible' in thirdParty) || !thirdParty['isVisible']) {
+        if (!(this.onWhiteList(origin, target))) {
+          // show third party
+          thirdParty['isVisible'] = true;
+          isNewThirdParty = true;
+        } else {
+          // hide third party
+          thirdParty['isVisible'] = false;
+        }
+      }
+      if (thirdParty['isVisible'] && isNewFirstParty) {
+        // an existing visible third party links to a new first party
+        thirdParty['firstPartyHostnames'].forEach((firstPartyHostname) => {
+          this.addFirstPartyLink(firstPartyHostname, target);
+        });
```

Comment:
However also missing is the:
```
this.addFirstPartyRelationship(origin, target);
```
This can happen where this code is or somewhere inside the following if statement:
```
    if (!('thirdPartyHostnames' in  firstParty)
|| !firstParty['thirdPartyHostnames'].includes(target)) {
```

The for loop could also ignore the origin too perhaps but that also makes following the logic complex.